src/Extension/EscaperExtension.php

```diff
@@ -84,59 +82,72 @@ public function setDefaultStrategy($defaultStrategy): void
      * @param string $name The template name
      *
      * @return string|false The default strategy to use for the template
+     *
+     * @deprecated since Twig 3.10
      */
     public function getDefaultStrategy(string $name)
     {
-        // disable string callables to avoid calling a function named html or js,
-        // or any other upcoming escaping strategy
-        if (!\is_string($this->defaultStrategy) && false !== $this->defaultStrategy) {
-            return \call_user_func($this->defaultStrategy, $name);
-        }
+        trigger_deprecation('twig/twig', '3.10', 'The "%s()" method is deprecated, use the "Twig\Runtime\EscaperRuntime::getDefaultStrategy()" method instead.', __METHOD__);
 
-        return $this->defaultStrategy;
+        return $this->escaper->getDefaultStrategy($name);
     }
 
     /**
      * Defines a new escaper to be used via the escape filter.
      *
-     * @param string   $strategy The strategy name that should be used as a strategy in the escape call
-     * @param callable $callable A valid PHP callable
+     * @param string                        $strategy The strategy name that should be used as a strategy in the escape call
+     * @param callable(Environment, string) $callable A valid PHP callable
+     *
+     * @deprecated since Twig 3.10
      */
     public function setEscaper($strategy, callable $callable)
     {
+        trigger_deprecation('twig/twig', '3.10', 'The "%s()" method is deprecated, use the "Twig\Runtime\EscaperRuntime::setEscaper()" method instead (be warned that Environment is not passed anymore to the callable).', __METHOD__);
+
+        if (!isset($this->environment)) {
+            throw new \LogicException('You must call setEnvironment() before calling setEscaper().');
+        }
+
         $this->escapers[$strategy] = $callable;
+        $callable = function ($string, $charset) use ($callable) {
+            return $callable($this->environment, $string, $charset);
+        };
+
+        $this->escaper->setEscaper($strategy, $callable);
     }
 
     /**
      * Gets all defined escapers.
      *
-     * @return callable[] An array of escapers
+     * @return array<callable(Environment, string)> An array of escapers
```

Comment:
those callables also take the charset as third argument

Reply:
The code and tests were wrong. Here, we should support the legacy escapers signature.

Reply:
See #4071 for the fix.